### PR TITLE
added flatten filter

### PR DESCRIPTION
**Pull request by bcoca**
##### SUMMARY

filter that flattens ... much easier to use than the lookup and has a 'level' interface

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

filters/flatten
##### ANSIBLE VERSION

```
2.5
```

### CODE REVIEW

File: docs/docsite/rst/playbooks_filters.rst [hunk 1]
```diff
@@ -116,6 +116,19 @@ To get the maximum value from a list of numbers::
 
     {{ [3, 4, 2] | max }}
 
+.. versionadded:: 2.5
+
+Flatten a list (same thing the `flatten` lookup does)::
+
+    {{ [3, [4, 2] ]|flatten }}
+
+Flatten only the first level of a list (aking to the `items` lookup)::
```
Comment (mgedmin) [inline docs/docsite/rst/playbooks_filters.rst]: Typo?  aking -> akin

File: lib/ansible/plugins/filter/core.py [hunk 2]
```diff
@@ -465,13 +465,38 @@ def b64decode(string):
...
+            if levels:
+                levels = int(levels) - 1
+                if levels >= 0:
+                    ret.extend(flatten(element, levels=levels))
+                else:
+                    ret.append(element)
+            elif levels is None:
+                ret.extend(flatten(element))
+            else:
+                break
```
Comment (mgedmin) [inline lib/ansible/plugins/filter/core.py]: This seems wrong to me.

Consider `levels=1`.  You decrement that to `levels=0`, then the `if levels >= 0:` branch gets taken and you recursively call `flatten(element, levels=0)`.  Here you find a nested subelement that is a list, so you check `if levels:` which is False, then you check `if levels is None`, which is also False, so you break, truncating `mylist` at the 1st item that happens to be a list.

Comment (bcoca) [reply lib/ansible/plugins/filter/core.py]: hmm, not sure how to check that initial input in levels is non 0

Comment (bcoca) [reply lib/ansible/plugins/filter/core.py]: @mgedmin should be fixed now

**Pull request merged**